bolt-sidecar/src/builder/fallback/engine_hinter.rs

```diff
@@ -0,0 +1,259 @@
+            }
+        }
+    }
+
+    /// Yield the next hint from the engine API by calling `engine_newPayloadV3`
+    /// and parsing the response to extract the hint.
+    ///
+    /// Returns Ok([EngineApiHint::ValidPayload]) if the payload is valid.
+    async fn next_hint(
+        &self,
+        exec_payload: ExecutionPayloadV3,
+        ctx: &EngineHinterContext,
+    ) -> Result<EngineApiHint, BuilderError> {
+        let payload_status = self
+            .engine_client
+            .new_payload_v3(
+                exec_payload,
+                ctx.blob_versioned_hashes.clone(),
+                ctx.parent_beacon_block_root,
+            )
+            .await?;
+
+        let validation_error = match payload_status.status {
+            PayloadStatusEnum::Valid => return Ok(EngineApiHint::ValidPayload),
+            PayloadStatusEnum::Invalid { validation_error } => validation_error,
+            PayloadStatusEnum::Syncing | PayloadStatusEnum::Accepted => {
+                error!(status = ?payload_status.status, "Unexpected payload status from engine API");
+                return Err(BuilderError::UnexpectedPayloadStatus(payload_status.status))
+            }
+        };
+
+        // Parse the hint from the engine API response, based on the EL client code
+        let Some(hint) = parse_hint_from_engine_response(ctx.el_client_code, &validation_error)?
+        else {
+            return Err(BuilderError::FailedToParseHintsFromEngine);
+        };
+
+        Ok(hint)
+    }
+}
+
+/// Engine API hint values that can be fetched from the engine API
+/// to complete the sealed block. These hints are used to fill in
+/// missing values in the block header.
+#[derive(Debug, Copy, Clone)]
+#[allow(clippy::large_enum_variant)]
+pub enum EngineApiHint {
+    BlockHash(B256),
+    GasUsed(u64),
+    StateRoot(B256),
+    ReceiptsRoot(B256),
+    LogsBloom(Bloom),
+    ValidPayload,
+}
+
+/// The collection of hints that can be fetched from the engine API
+/// via the [EngineHinter] to complete the sealed block.
+///
+/// When a field is `None`, we set it to its default value in the [ExecutionPayload]
+/// and try to get the hint from the engine API response to fill its value.
+#[derive(Debug, Default)]
+pub struct Hints {
+    pub gas_used: Option<u64>,
+    pub receipts_root: Option<B256>,
+    pub logs_bloom: Option<Bloom>,
+    pub state_root: Option<B256>,
+    pub block_hash: Option<B256>,
+}
+
+impl Hints {
+    /// Populate the new hint value in the context.
+    pub fn populate_new(&mut self, hint: EngineApiHint) {
+        match hint {
+            EngineApiHint::ValidPayload => { /* No-op */ }
+
+            // If we receive a block hash hint, set it and keep it for the next one.
+            // This should not happen, but in case it does, it doesn't break the flow.
+            EngineApiHint::BlockHash(hash) => self.block_hash = Some(hash),
+
+            // For regular hint types, set the value and reset the block hash
```

Comment:
Why do we reset the block hash hint here every time?

Reply:
It really depends on how the EL validation is structured, but me and @thedevbirb found that sometimes Geth would return a mismatching hash error after populating some extra field (e.g. state root) which means we can't get more information out of it (as it kept returning `mismatching block hash` and no other hints). So when this happens we set the block hash to the one hinted by Geth and then proceed with the next hints. 

As the comment mentions, this shouldn't happen and it's more of an edge case